crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```diff
@@ -1,24 +1,273 @@
+
+    padw mem_loadw.MESSAGE movup.4
+    # => [num_required_signatures, MESSAGE]
+
+    # Counter `i` starts at `num_required_signatures` and counts down to 0
+    # => [i, MESSAGE]
+
+    # Loop through required signers and verify signatures.
+    # ---------------------------------------------------------------------------------------------
+
+    dup neq.0
+    while.true
+        # => [i, MESSAGE]
+
+        # Fetch public key from storage map.
+        # -----------------------------------------------------------------------------------------
+
+        dup sub.1 push.0.0.0 push.PUBLIC_KEYS_MAP_SLOT
+        # => [owner_key_slot, [0, 0, 0, i-1], i, MESSAGE]
+
+        exec.account::get_map_item
+        # => [OWNER_PUB_KEY, i, MESSAGE]
+
+        mem_storew.TEMP_PK
+        # => [OWNER_PUB_KEY, i, MESSAGE]
+
+        # Check if signature exists for this signer.
+        # -----------------------------------------------------------------------------------------
+
+        movup.4 movdn.8 swapw dupw movdnw.2
+        # => [MESSAGE, OWNER_PUB_KEY, MESSAGE, i]
+
+        hmerge
+        # => [SIG_KEY, MESSAGE, i]
+
+        adv.has_mapkey
+        # => [SIG_KEY, MESSAGE, i]
+
+        adv_push.1
+        # => [has_signature, SIG_KEY, MESSAGE, i]
+
+        # if h(msg, pk) => msg exists in AdviceMap check the signature
+        if.true
+            dropw dupw
+            # => [MSG, MSG, i]
+
+            # Verify the signature against the public key and message.
+            # -----------------------------------------------------------------------------------------
+
+            padw mem_loadw.TEMP_PK
+            # => [PK, MSG, MSG, i]
+
+            swapw
+            # => [MSG, PK, MSG, i]
+
+            # Emit the authentication request event that pushes a signature for the message to the advice stack.
+            emit.AUTH_REQUEST
+
+            swapw
+            # => [PK, MSG, MSG, i ]
+
+            exec.rpo_falcon512::verify
+            # => [MSG, i]
+
+            mem_load.SUCCESSFUL_VERIFICATIONS
+            add.1
+            mem_store.SUCCESSFUL_VERIFICATIONS
+            # => [MSG, i]
+
+        else
+            dropw
+            # => [MSG, i]
+        end
+        # => [MSG, i]
+
+        movup.4
+        # => [i, MSG]
+
+        # Decrement counter and check if we should continue looping
+        sub.1
+        # => [i-1, MSG]
+
+        dup neq.0
+        # => [should_continue, i-1, MSG]
+    end
+    # => [i-1, MSG]
+
+    # Check number of successful signature verifications
+    # -----------------------------------------------------------------------------------------
+
+    drop
+    # => [MSG]
+
+    mem_load.SUCCESSFUL_VERIFICATIONS dup
+    # => [num_verifications, num_verifications, MSG]
+
+    eq.0 swap
+    # => [num_verifications, is_zero_verifications, MSG]
+
+    mem_load.THRESHOLD
+    # => [threshold, num_verifications, is_zero_verifications, MSG]
+
+    gte not or
+    # => [is_unauthorized, MSG]
```

Comment:
do we need the condition `is_zero`? I think the case of `num_verifications < threshold` covers that